### ISSUE TITLE
Dependency Version Update Breaks Expected Behavior

### ISSUE THREAD

**Issue by everythingfunctional** [fortran-lang/fpm#870]
### Description

The latest release (0.8.0) breaks some of my projects, as it changes the way the dependency resolution is done. The new behavior introduced in #843 causes the versions specified in the main project `fpm.toml` file not to be respected.

### Expected Behaviour

The versions of dependencies specified in the `fpm.toml` file should be respected as the intended version.

### Version of fpm

0.8.0

### Platform and Architecture

Linux x86 (but all will observe this behavior)

### Additional Information

The intended behavior was that the first time a dependency is encountered (in a breadth first search of the dependency tree), that is the version that should be used. This has the effect that all versions specified in the `fpm.toml` file of the main project take precedence. The behavior introduced in #843 seems to have inverted this. I.e. the version specified in a dependency gets recorded in the cache, it gets detected as a version update, and it uses that version, meaning the version specified in the main project gets overridden.

**Comment by perazz** [comment]
I am sorry @everythingfunctional that this change affected your projects, my bad this feature was not evident from the code and did not appear in the documentation. Do you have a minimal reproducer (i.e., an fpm.toml with two cascaded dependencies) I could add as a test? I will try to work on this across the Easter weekend.

**Comment by everythingfunctional** [comment]
Not immediately, but Something like the following should demonstrate

Main package:
```
pack_a = { ..., tag = "v0.3.0"
pack_b = { ..., tag = "v0.2.0"
pack_c = { ..., tag = "v0.1.0"
```

pack_a:
```
pack_c = { ...., tag = "v0.1.0" }
```

pack_b:
```
pack_c = { ..., tag = "v0.0.1" }
```

Note how pack_b hasn't been updated to the latest version of pack_c, but the main package and pack_a both likely need the latest version. The current behavior causes the version from pack_b to be used though.

**Comment by everythingfunctional** [comment]
> I believe that the initial implementation of the upcoming version constraint resolution feature will replace this logic, as as far as I understand the latest/most recent version among the subset will be chosen, at least until more complex rules will be available

For packages in the official registry I think this will be the case, but I think the current logic for git dependencies will likely need to be preserved, as it won't ask the registry about them.

**Comment by perazz** [comment]
Thank you @everythingfunctional on a much needed check. Closed by #871, #873, #875
